pkg/cli/cmd_helpers.go

```diff
@@ -17,69 +17,275 @@ limitations under the License.
 // or used with the help flag.
 func cmdErr(cmd *cobra.Command, err error) {
 	cmd.Long = fmt.Sprintf("%s\nNote: %v", cmd.Long, err)
 	cmd.RunE = errCmdFunc(err)
 }
 
-// cmdErrNoHelp calls cmdErr(cmd, err) then turns cmd's usage off.
-func cmdErrNoHelp(cmd *cobra.Command, err error) {
-	cmdErr(cmd, err)
-	cmd.SilenceUsage = true
-}
-
 // errCmdFunc returns a cobra RunE function that returns the provided error
 func errCmdFunc(err error) func(*cobra.Command, []string) error {
 	return func(*cobra.Command, []string) error {
 		return err
 	}
 }
 
-// preRunECmdFunc returns a cobra PreRunE function that loads the configuration file
-// and injects it into the subcommand
-func preRunECmdFunc(subcmd plugin.Subcommand, cfg store.Store, msg string) func(*cobra.Command, []string) error {
+// filterSubcommands returns a list of plugin keys and subcommands from a filtered list of resolved plugins.
+func (c CLI) filterSubcommands(
+	filter func(plugin.Plugin) bool,
+	extract func(plugin.Plugin) plugin.Subcommand,
+) ([]string, *[]plugin.Subcommand) {
+	// Unbundle plugins
+	plugins := make([]plugin.Plugin, 0, len(c.resolvedPlugins))
+	for _, p := range c.resolvedPlugins {
+		if bundle, isBundle := p.(plugin.Bundle); isBundle {
+			plugins = append(plugins, bundle.Plugins()...)
+		} else {
+			plugins = append(plugins, p)
+		}
+	}
+
+	pluginKeys := make([]string, 0, len(plugins))
+	subcommands := make([]plugin.Subcommand, 0, len(plugins))
+	for _, p := range plugins {
+		if filter(p) {
+			pluginKeys = append(pluginKeys, plugin.KeyFor(p))
+			subcommands = append(subcommands, extract(p))
+		}
+	}
+	return pluginKeys, &subcommands
+}
+
+// initializationMethods
+func (c CLI) initializationMethods(cmd *cobra.Command, subcommands *[]plugin.Subcommand) *resourceOptions {
+	// Update metadata method.
+	meta := plugin.SubcommandMetadata{
+		Description: cmd.Long,
+		Examples:    cmd.Example,
+	}
+	for _, subcommand := range *subcommands {
+		if subcmd, updatesMetadata := subcommand.(plugin.UpdatesMetadata); updatesMetadata {
+			subcmd.UpdateMetadata(c.metadata(), &meta)
+		}
+	}
+	cmd.Long = meta.Description
+	cmd.Example = meta.Examples
+
+	// Before binding specific plugin flags, bind common ones
+	requiresResource := false
+	for _, subcommand := range *subcommands {
+		if _, requiresResource = subcommand.(plugin.RequiresResource); requiresResource {
+			break
+		}
+	}
+	var options *resourceOptions
+	if requiresResource {
+		options = bindResourceFlags(cmd.Flags())
+	}
+
+	// Bind flags method.
+	for _, subcommand := range *subcommands {
+		if subcmd, hasFlags := subcommand.(plugin.HasFlags); hasFlags {
+			subcmd.BindFlags(cmd.Flags())
+		}
+	}
+
+	return options
+}
+
+// executionMethodsFuncs returns cobra RunE functions for PreRunE, RunE, PostRunE cobra hooks.
+func (c CLI) executionMethodsFuncs(
+	pluginKeys []string,
+	subcommands *[]plugin.Subcommand,
+	options *resourceOptions,
+	msg string,
+) (
+	func(*cobra.Command, []string) error,
```

Comment:
nit: return a struct or something for this -- more than two return values is pushing it in terms of readabilty